File: sw/sim/flash.c

/*
 * Copyright 2021 Nicolas Maltais
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <sim/flash.h>
#include <sim/memory.h>
#include <sys/flash.h>

#include <core/trace.h>

#include <stddef.h>
#include <stdbool.h>
#include <memory.h>
#include <pthread.h>

#define READ_BUFFER_SIZE 8192
#define ERASE_BYTE 0xff

#define STATUS_WREN_MASK 0x02
#define STATUS_BUSY_MASK 0x01

#define PAGE_MASK 0xff  // page size = 256

enum {
    INSTRUCTION_READ = 0x03,
    INSTRUCTION_WRITE = 0x02,
    INSTRUCTION_ERASE_4KB = 0x20,
    INSTRUCTION_ERASE_32KB = 0x52,
    INSTRUCTION_ERASE_64KB = 0xd8,
    INSTRUCTION_ERASE_ALL = 0x60,
    INSTRUCTION_WRITE_ENABLE = 0x06,
    INSTRUCTION_WRITE_DISABLE = 0x04,
    INSTRUCTION_READ_STATUS = 0x05,
    INSTRUCTION_WRITE_STATUS = 0x01,
    INSTRUCTION_RESET_ENABLE = 0x66,
    INSTRUCTION_RESET = 0x99,
    INSTRUCTION_POWER_DOWN_ENABLE = 0xb9,
    INSTRUCTION_POWER_DOWN_DISABLE = 0xab,
    INSTRUCTION_MANUF_ID = 0x9f,
};

static const uint8_t MANUFACTURER_ID[] = {0x1f, 0x85, 0x01};

sim_mem_t* flash;
pthread_mutex_t flash_mutex;

static struct {
    uint8_t status;
    uint8_t instr;
    flash_t address;
    size_t pos;
    bool writing;
    bool reset_enabled;
    bool power_down;
} spi_flash;

void sim_flash_init(void) {
    flash = sim_mem_init(SYS_FLASH_SIZE, ERASE_BYTE);
}

void sim_flash_free(void) {
    pthread_mutex_lock(&flash_mutex);
    sim_mem_free(flash);
    flash = 0;
    pthread_mutex_unlock(&flash_mutex);
}

void sim_flash_load(const char* filename) {
    pthread_mutex_lock(&flash_mutex);
    sim_mem_load(flash, filename);
    pthread_mutex_unlock(&flash_mutex);
    // in simulation, flash always starts at address 0
    sys_flash_set_offset(0);
}

void sim_flash_save(void) {
    pthread_mutex_lock(&flash_mutex);
    sim_mem_save(flash);
    pthread_mutex_unlock(&flash_mutex);
}

void sim_flash_spi_transceive(size_t length, uint8_t data[static length]) {
    pthread_mutex_lock(&flash_mutex);
    if (!flash) {
        return;
    }

    for (size_t i = 0; i < length; ++i) {
        uint8_t b = data[i];
        size_t pos = spi_flash.pos;

        if (spi_flash.instr == 0 && pos == 0) {
            if (spi_flash.power_down && b != INSTRUCTION_POWER_DOWN_DISABLE) {
                // flash is in deep power-down, all commands ignored except wakeup.
                trace("instruction 0x%02x ignored in power-down mode.", b);
                return;
            }
            // set new instruction
            spi_flash.instr = b;
        }

        switch (spi_flash.instr) {
            case INSTRUCTION_READ: {
                if (pos == 1) {
                    spi_flash.address = b << 16;
                } else if (pos == 2) {
                    spi_flash.address |= b << 8;
                } else if (pos == 3) {
                    spi_flash.address |= b;
                    spi_flash.address %= SYS_FLASH_SIZE;
                } else if (pos >= 4) {
                    data[i] = flash->data[spi_flash.address];
                    spi_flash.address = (spi_flash.address + 1) % SYS_FLASH_SIZE;
                }
                break;
            }
            case INSTRUCTION_WRITE: {
                if (pos == 0) {
                    spi_flash.writing = (spi_flash.status & STATUS_WREN_MASK) != 0;
                    spi_flash.status &= ~STATUS_WREN_MASK;
                    spi_flash.status |= STATUS_BUSY_MASK;
                } else if (pos == 1) {
                    spi_flash.address = b << 16;
                } else if (pos == 2) {
                    spi_flash.address |= b << 8;
                } else if (pos == 3) {
                    spi_flash.address |= b;
                    spi_flash.address %= SYS_FLASH_SIZE;
                } else if (spi_flash.writing) {
                    // data AND with existing data like the real flash device.
                    flash->data[spi_flash.address] &= b;
                    spi_flash.address = (spi_flash.address & ~PAGE_MASK) |
                                        ((spi_flash.address + 1) & PAGE_MASK);
                }
                break;
            }
            case INSTRUCTION_ERASE_4KB:
            case INSTRUCTION_ERASE_32KB:
            case INSTRUCTION_ERASE_64KB: {
                if (pos == 0) {
                    spi_flash.writing = (spi_flash.status & STATUS_WREN_MASK) != 0;
                    spi_flash.status &= ~STATUS_WREN_MASK;
                    spi_flash.status |= STATUS_BUSY_MASK;
                } else if (pos == 1) {
                    spi_flash.address = b << 16;
                } else if (pos == 2) {
                    spi_flash.address |= b << 8;
                } else if (pos == 3) {
                    spi_flash.address |= b;
                    spi_flash.address %= SYS_FLASH_SIZE;

                    // erase the block
                    size_t size;
                    if (spi_flash.instr == INSTRUCTION_ERASE_4KB) {
                        size = 0x1000;
                    } else if (spi_flash.instr == INSTRUCTION_ERASE_32KB) {
                        size = 0x8000;
                    } else {
                        size = 0x10000;
                    }
                    spi_flash.address = spi_flash.address & ~(size - 1);
                    memset(&flash->data[spi_flash.address], ERASE_BYTE, size);
                }
                break;
            }
            case INSTRUCTION_ERASE_ALL: {
                if (pos == 0) {
                    if (spi_flash.status == STATUS_WREN_MASK) {
                        memset(flash->data, ERASE_BYTE, SYS_FLASH_SIZE);
                    }
                    spi_flash.status &= ~STATUS_WREN_MASK;
                    spi_flash.status |= STATUS_BUSY_MASK;
                }
                break;
            }
            case INSTRUCTION_WRITE_ENABLE: {
                spi_flash.status |= STATUS_WREN_MASK;
                break;
            }
            case INSTRUCTION_WRITE_DISABLE: {
                spi_flash.status &= ~STATUS_WREN_MASK;
                break;
            }
            case INSTRUCTION_READ_STATUS: {
                if (pos == 1) {
                    data[i] = spi_flash.status;
                    // after write, status will read as busy once, then ready.
                    spi_flash.status &= ~STATUS_BUSY_MASK;
                }
                break;
            }
            case INSTRUCTION_WRITE_STATUS: {
                if (pos == 1) {
                    spi_flash.status = b & 0xf4;
                }
                break;
            }
            case INSTRUCTION_RESET_ENABLE: {
                spi_flash.reset_enabled = true;
                break;
            }
            case INSTRUCTION_RESET: {
                if (pos == 0 && spi_flash.reset_enabled) {
                    spi_flash.status = 0;
                    spi_flash.reset_enabled = false;
                }
                break;
            }
            case INSTRUCTION_POWER_DOWN_ENABLE: {
                spi_flash.power_down = true;
                break;
            }
            case INSTRUCTION_POWER_DOWN_DISABLE: {
                spi_flash.power_down = false;
                break;
            }
            case INSTRUCTION_MANUF_ID: {
                if (pos > 0 && pos <= sizeof MANUFACTURER_ID) {
                    data[i] = MANUFACTURER_ID[spi_flash.pos - 1];
                }
                break;
            }
            default:
                if (pos == 0) {
                    trace("unsupported flash instruction 0x%02x", spi_flash.instr);
                }
                break;
        }

        ++spi_flash.pos;
    }

    pthread_mutex_unlock(&flash_mutex);
}

void sim_flash_spi_reset(void) {
    spi_flash.instr = 0;
    spi_flash.address = 0;
    spi_flash.pos = 0;
    spi_flash.writing = false;
}

File: sw/utils/prog/flash.py
#  Copyright 2022 Nicolas Maltais
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

# This module is used to read, write & erase the flash chip on the game console
# The flash chip model number is AT25SF081B.
# Datasheet: https://www.dialog-semiconductor.com/sites/default/files/at25sf081b.pdf
# Flash devices support SFDP which could allow this driver to work for any flash device, but parsing
# the SFDP table is more tedious than just hardcoding the values for a specific chip.

import time
from dataclasses import dataclass
from pathlib import Path
from typing import Union

from prog.comm import ProgError
from prog.memory import MemoryDriver, MemoryLocal
from prog.spi import SpiInterface, SpiPeripheral
from utils import ProgressCallback, process_bitarray_sequences
from bitarray import bitarray

FLASH_SIZE = 1048576
# flash page size in bytes
PAGE_SIZE = 256
ADDRESS_BYTES = 3

# driver will abort the operation if the flash chip manufacturer & device ID
# does not match this value (corresponds to the 3 bytes returned by command 0x9f).
MANUFACTURER_ID = 0x01851f


@dataclass(frozen=True)
class EraseInstruction:
    opcode: int
    typ_time: float


SMALLEST_BLOCK_SIZE = 4096

# instruction set
INSTR_READ = 0x03
INSTR_WRITE = 0x02
INSTR_ERASE = {
    SMALLEST_BLOCK_SIZE: EraseInstruction(0x20, 0.06),
    32768: EraseInstruction(0x52, 0.135),
    65536: EraseInstruction(0xd8, 0.220),
    FLASH_SIZE: EraseInstruction(0x60, 3),
}
INSTR_WRITE_EN = 0x06
INSTR_READ_STATUS = 0x05
INSTR_RESET_ENABLE = 0x66
INSTR_RESET = 0x99
INSTR_MANUF_ID = 0x9f


class FlashDriver(MemoryDriver):
    """Class used to read & write to the flash chip."""
    spi: SpiInterface

    def __init__(self, spi: SpiInterface):
        spi.peripheral = SpiPeripheral.FLASH
        self.spi = spi

    def reset(self) -> None:
        """Reset device and verify manufacturer ID."""
        self._wait_ready()
        self.spi.transceive([INSTR_RESET_ENABLE])
        self.spi.transceive([INSTR_RESET])
        id_data = self.spi.transceive([INSTR_MANUF_ID, 0, 0, 0])
        id_num = id_data[1] | id_data[2] << 8 | id_data[3] << 16
        if id_num != MANUFACTURER_ID:
            raise ProgError("Flash device manufacturer and device ID do not match predefined value")

    def get_size(self) -> int:
        return FLASH_SIZE

    def requires_block_erase(self) -> bool:
        return True

    def get_smallest_erase_size(self) -> int:
        return SMALLEST_BLOCK_SIZE

    def get_block_count(self) -> int:
        return FLASH_SIZE // SMALLEST_BLOCK_SIZE

    def read(self, address: int, count: int, progress: ProgressCallback) -> bytes:
        self._wait_ready()
        command = bytearray([INSTR_READ])
        command += address.to_bytes(3, "big", signed=False)
        # insert dummy MOSI bytes
        for i in range(count):
            command.append(0)
        received = self.spi.transceive(command, progress=lambda c, t: progress(c - 4, t - 4))
        return received[4:]

    def write(self, address: int, data: bytes, progress: ProgressCallback) -> None:
        """Write bytes to flash at an address.
        All bytes written must be in erased state (0xff) beforehand."""
        if address % PAGE_SIZE != 0 or len(data) % PAGE_SIZE != 0:
            raise ValueError("Write address and data size must be page aligned!")
        pos = 0
        self._wait_ready()
        while pos < len(data):
            self._write_enable()
            # write data for one page
            command = bytearray([INSTR_WRITE])
            command += address.to_bytes(ADDRESS_BYTES, "big", signed=False)
            for i in range(PAGE_SIZE):
                command.append(data[pos])
                pos += 1
            self.spi.transceive(command)
            self._wait_ready()
            progress(pos, len(data))
            address += PAGE_SIZE

    def erase_blocks(self, blocks: bitarray, progress: ProgressCallback) -> None:
        # for each sequence, erase with the largest possible block since it's faster.
        # however the erase block start address must be aligned to the block size.
        # estimate progress using typical erase times given in datasheet.
        self._wait_ready()

        def erase_process(start: int, end: int, _) -> None:
            nonlocal erased
            while start != end:
                for block_size in reversed(INSTR_ERASE):
                    if start % block_size == 0 and end - start >= block_size:
                        # this is the largest block erase possible from current position
                        self._write_enable()
                        instr = INSTR_ERASE[block_size]
                        command = bytearray([instr.opcode])
                        if block_size != FLASH_SIZE:
                            # erase instructions require an address, except chip erase
                            command += start.to_bytes(ADDRESS_BYTES, "big", signed=False)
                        self.spi.transceive(command)
                        # estimate progress using typical datasheet erase time
                        # blocks at last byte if not done after that time.
                        start_time = time.time()
                        while not self._is_ready():
                            elapsed = time.time() - start_time
                            block_erased_est = min(block_size - 1,
                                                   round(block_size * elapsed / instr.typ_time))
                            progress(erased + block_erased_est, count)
                        erased += block_size
                        progress(erased, count)
                        start += block_size
                        break

        erased = 0
        count = blocks.count() * SMALLEST_BLOCK_SIZE
        process_bitarray_sequences(self, blocks, "", erase_process, verbose=False)

    def erase(self, progress: ProgressCallback) -> None:
        """Erase whole device."""
        all_blocks = bitarray(self.get_block_count())
        all_blocks.setall(True)
        self.erase_blocks(all_blocks, progress)

    def _write_enable(self) -> None:
        self.spi.transceive([INSTR_WRITE_EN])

    def _is_ready(self) -> bool:
        """Returns true if flash device is ready."""
        received = self.spi.transceive([INSTR_READ_STATUS, 0])
        return not (received[1] & 0x1)

    def _wait_ready(self) -> None:
        """Wait until flash device is ready."""
        while not self._is_ready():
            pass

    @staticmethod
    def local(filename: Union[str, Path]) -> MemoryLocal:
        return MemoryLocal(filename, FLASH_SIZE, 0xff, SMALLEST_BLOCK_SIZE, True)
